text/2018-04-08-global-and-local-pinning.md

```diff
@@ -0,0 +1,247 @@
+
+1. Users might mistakenly expect `Handle::clone()` to register a brand-new handle rather than returning a reference to the same handle. The method can be dangerously misleading.
+
+2. Handles are not `Send`. This is an annoying problem and it's not obvious how to cheat around it correctly when you really have to.
+
+3. Handles are not `Sync`. Sometimes it'd be nice to call `pin()` without registering a handle for each thread. In `SkipMap`, we need to pin inside `self.collector`, but creating a handle per-thread *and* per-instance is difficult. In `elfmalloc` there are situations where having to register handles is a nuisance.
+
+4. `epoch::pin()` will panic if the thread-local handle is destroyed. Fortunately, `LocalKey::try_with` will be soon stabilized in Rust 1.26. But how do we work around the panics then - what should we do exactly in case the handle is destroyed?
+
+# Detailed design
+
+### Global pinning
+
+Let's introduce something I've been referring to as *anonymous pinning* in the past, and has been briefly explained in [this comment](https://github.com/crossbeam-rs/crossbeam-epoch/pull/21#discussion_r144353198).
+
+The idea is to have two atomic counters in the `Collector` (more concretely, in the `Global` struct). They count how many `Guards` are currently pinned in the global epoch and the epoch preceding it:
+
+```rust
+pin_counters: [AtomicUsize; 2],
+```
+
+We introduce a new method `Collector::pin()`, which is just like `Handle::pin()`:
+
+```rust
+impl Collector {
+    fn pin(&self) -> Guard;
+}
+```
+
+We'll have to distinguish between guards pinned globally and locally. Since we're turning `Guard` into an `enum`, let's also add a special case for the unprotected guard:
+
+```rust
+enum Guard {
+    Global(*const Global),
+    Local(*const Local),
+    Unprotected,
+}
+```
+
+Now, the `Global::pin()` method might be implemented like this:
+
+```rust
+fn pin(&self) -> Guard {
+    // 0 if the epoch is even, or 1 if the epoch is odd.
+    let parity = self.epoch.load(Ordering::Relaxed).parity();
+
+    self.pin_counters[parity].fetch_add(1, Ordering::Relaxed);
+    atomic::fence(Ordering::SeqCst);
+
+    Guard::Global(self)
+}
+```
+
+Of course, unpinning will simply decrement the counter.
+
+**@jeehoonkang** [raised a concern](https://github.com/crossbeam-rs/crossbeam-epoch/pull/21#discussion_r144719080) with correctness of this method of pinning.
+
+Consider this. What happens if we read a stale value of `self.epoch` and increment a counter just after the epoch advanced one step forward, and execute the fence just before the epoch is advanced one step forward again? That means we're effectively two epochs behind, while other threads might think we're zero epochs behind.
+
+I think the problem is easily fixable by changing when a bag becomes expired. Rather than defining a bag is expired if it's at least 2 epochs old, we should say it's expired if it's at least 3 epochs old. That's it.
```

Comment:
Please forgive my ignorance, but as I understand, the benefit of using parity over a single counter is that with parity the epoch can still be advanced by 1 when there is a global pin present, right? If a larger counter is used, say `mod 4`, and then the epoch age required for destruction extended to 5, can the epoch advance by up to 3 with a global pin present? I'm not really proposing anything like that, but I'm curious about the relationship between the number of counters and the minimal free-able epoch age.

Reply:
> Please forgive my ignorance, but as I understand, the benefit of using parity over a single counter is that with parity the epoch can still be advanced by 1 when there is a global pin present, right?